site/assignments/tic-tac-toe/asn3.py

```diff
@@ -0,0 +1,360 @@
+    right cell, can be seen below.
+
+        [[' ', ' ', 'O'],
+         [' ', ' ', ' '],
+         ['X', ' ', ' ']]
+
+    :param size: Size of the game board to be returned.
+    :type size: Integer
+    :return: An empty tic-tac-toe game board with all cells set blank.
+    :rtype: A list of three lists containing three characters (strings) each.
+    """
+
+
+
+assert [[" "]] == setup_game(1)
+assert [[" ", " "], [" ", " "]] == setup_game(2)
+assert [[" ", " ", " "], [" ", " ", " "], [" ", " ", " "]] == setup_game(3)
+assert [[" ", " ", " ", " "], [" ", " ", " ", " "], [" ", " ", " ", " "], [" ", " ", " ", " "]] == setup_game(4)
+test_board = setup_game(3)
+assert not (test_board[0] is test_board[1])
+assert not (test_board[0] is test_board[2])
+assert not (test_board[1] is test_board[2])
+
+def parse_move(string: str) -> tuple:
+    """
+    Parse a move from the string provided as a parameter and return it as a tuple of integers of size 2. First index is
+    the column (x), second index is the row (y). Input is expected to be in the form of as "x, y" --- the comma is
+    necessary.
+
+        x --- the x/column of the move
+        y --- the y/row of the move
+
+    The top left corner of the game board is (0, 0).
+
+    If more than two comma separated values are provided, only the first 2 are returned.
+
+    :param string: String of a move to be parsed.
+    :type string: String
+    :return: The move, as specified as a column, row pair.
+    :rtype: A tuple of integers of size 2 (int(x), int(y))
+    """
+
+
+
+assert (0, 0) == parse_move("0, 0")
+assert (2, 1) == parse_move("2,1")
+assert (10, 11) == parse_move("10, 11, 12, 13")
+
+def is_move_valid(move: tuple, board: list) -> bool:
+    """
+    Checks if the provided move is valid for the current game board. A move is valid if (a) the cell it refers to is
+    unoccupied, and (b) it is within the game board.
+
+    :param move: A tuple with the player's move's x and y location.
+    :type move: A tuple of integers of size 2 (int(x), int(y)).
+    :param board: The current game board.
+    :type board: A list of n lists containing n characters (strings) each.
+    :return: True if the move is considered valid, False otherwise.
+    :rtype: Boolean.
+    """
+
+
+
+test_board_3 = [[" ", " ", " "], ["O", "X", " "], [" ", " ", " "]]
+assert True == is_move_valid((0, 0), test_board_3)
+assert True == is_move_valid((2, 2), test_board_3)
+assert False == is_move_valid((-1, 0), test_board_3)
+assert False == is_move_valid((0, -1), test_board_3)
+assert False == is_move_valid((-1, -1), test_board_3)
+assert False == is_move_valid((3, 0), test_board_3)
+assert False == is_move_valid((0, 3), test_board_3)
+assert False == is_move_valid((3, 3), test_board_3)
+assert False == is_move_valid((1, 1), test_board_3)
+assert False == is_move_valid((0, 1), test_board_3)
+
+def apply_move(move: tuple, board: list, player: str) -> None:
```

Comment:
It currently has a side effect by altering the provided board. 

Thoughts on making it immutable by returning a new board?